src/expr/impl/src/scalar/external/iceberg.rs

```diff
@@ -0,0 +1,202 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module contains the expression for computing the iceberg partition value.
+//! spec ref: <https://iceberg.apache.org/spec/#partition-transforms>
+use std::fmt::Formatter;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use icelake::types::{create_transform_function, BoxedTransformFunction, Transform};
+use risingwave_common::array::{ArrayRef, DataChunk};
+use risingwave_common::ensure;
+use risingwave_common::row::OwnedRow;
+use risingwave_common::types::{DataType, Datum};
+use risingwave_expr::expr::BoxedExpression;
+use risingwave_expr::{build_function, ExprError, Result};
+
+pub struct IcebergTransform {
+    child: BoxedExpression,
+    transform: BoxedTransformFunction,
+    return_type: DataType,
+}
+
+impl std::fmt::Debug for IcebergTransform {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("IcebergTransform")
+            .field("child", &self.child)
+            .field("return_type", &self.return_type)
+            .finish()
+    }
+}
+
+#[async_trait::async_trait]
+impl risingwave_expr::expr::Expression for IcebergTransform {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, data_chunk: &DataChunk) -> Result<ArrayRef> {
+        // Get the child array
+        let array = self.child.eval(data_chunk).await?;
+        // Convert to arrow array
+        let arrow_array = array.as_ref().try_into().unwrap();
+        // Transform
+        let res_array = self.transform.transform(arrow_array).unwrap();
+        // Convert back to array ref and return it
+        Ok(Arc::new((&res_array).try_into().unwrap()))
+    }
+
+    async fn eval_row(&self, _row: &OwnedRow) -> Result<Datum> {
+        Err(ExprError::Internal(anyhow!(
+            "eval_row in iceberg_transform is not supported yet"
+        )))
+    }
+}
+
+#[build_function("iceberg_transform(varchar, any) -> any")]
+fn build(return_type: DataType, mut children: Vec<BoxedExpression>) -> Result<BoxedExpression> {
+    let transform_type = {
+        let datum = children[0].eval_const()?.unwrap();
+        let str = datum.as_utf8();
+        Transform::from_str(str).map_err(|_| ExprError::InvalidParam {
+            name: "transform type in icberg_transform",
+            reason: format!("Fail to parse {str} as iceberg transform type").into(),
+        })?
+    };
+
+    // Check type
+    match &transform_type {
+        Transform::Bucket(_) => IcebergTransform::check_bucket(&return_type, &children)?,
+        Transform::Truncate(_) => IcebergTransform::check_truncate(&return_type, &children)?,
+        Transform::Year | Transform::Month | Transform::Day | Transform::Hour => {
+            IcebergTransform::check_time(&return_type, &children)?
+        }
+        Transform::Identity | Transform::Void => {
```

Comment:
Why identity and void is illegal? Identity means user has calculated partition value for us, so I think it should be valid. `Void` is a little complex, if all partition fields are void, it should be treated as unpartitioned. But if only one field is partitioned, we should return null.

Reply:
For identity and void, we can directly represent them in the project as `InputRef` and const null.

Reply:
You mean do this in optimizer? Sounds reasonable to me, but please add some comment here.